src/Chainweb/Pact/SPV/Hyperlane.hs

```diff
@@ -0,0 +1,276 @@
+  , Just (TLitString metadata) <- M.lookup "metadata" om
+  , Just (TList validators _ _) <- M.lookup "validators" om
+  , Just (TLitInteger threshold) <- M.lookup "threshold" om
+  =
+    let
+      convert (TLitString v) = Just v
+      convert _ = Nothing
+    in verifySignatures message metadata (V.mapMaybe convert validators) (fromInteger threshold)
+
+  | Just (TObject o _) <- M.lookup "message" om
+  = encodeHyperlaneMessage o
+
+  | otherwise = throwError "Unknown hyperlane command"
+
+-- | Decodes Hyperlane binary message and metadata,
+-- verifies against the provided signatures using the provided threshold.
+verifySignatures :: Text -> Text -> V.Vector Text -> Int -> ExceptT Text IO (Object Name)
+verifySignatures hexMessage hexMetadata validators threshold = do
+  message <- case decodeHex hexMessage of
+          Right s -> pure s
+          Left e -> throwError $ Text.pack $ "Decoding of HyperlaneMessage failed: " ++ e
+
+  HyperlaneMessage{..} <- runGetS getHyperlaneMessage message
+
+  metadata <- case decodeHex hexMetadata of
+          Right s -> pure s
+          Left e -> throwError $ Text.pack $ "Decoding of Metadata failed: " ++ e
+
+  MessageIdMultisigIsmMetadata{..} <- runGetS getMessageIdMultisigIsmMetadata metadata
+
+  let
+    domainHash = getKeccak256Hash $ runPutS $ do
+      -- Corresponds to abi.encodePacked behaviour
+      putWord32be hmOriginDomain
+      putRawByteString mmimOriginMerkleTreeAddress
+      putRawByteString "HYPERLANE"
+
+  let messageId = getKeccak256Hash message
+
+  let
+    hash' = getKeccak256Hash $ runPutS $ do
+      -- Corresponds to abi.encodePacked behaviour
+      putRawByteString domainHash
+      putRawByteString mmimSignedCheckpointRoot
+      putWord32be mmimSignedCheckpointIndex
+      putRawByteString messageId
+  let
+    digest = keccak256 $ runPutS $ do
+      -- Corresponds to abi.encodePacked behaviour
+      putRawByteString ethereumHeader
+      putRawByteString hash'
+
+  addresses <- catMaybes <$> mapM (recoverAddress digest) mmimSignatures
+
+  when (length addresses < threshold) $
+    throwError $ Text.pack $ "The number of recovered addresses from the signatures is less than threshold: " ++ show threshold
+
+  binaryValidators <- forM validators $ \val -> case decodeHex val of
+          Right v -> pure v
+          Left e -> throwError $ "Failed to decode a validator (" <> val <> "):" <> Text.pack e
+
+  -- Requires that m-of-n validators verify a merkle root, and verifies a merkle proof of message against that root.
+  --
+  -- The signature addresses and validator addresses should be in the same order.
+  --
+  -- The original algorithm in hyperlane.
+  -- https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/v3/solidity/contracts/isms/multisig/AbstractMultisigIsm.sol#L67
+  let verificationAddresses = take threshold addresses
+  let verifyStep (_, vals) signer = case V.elemIndex signer vals of
+        Just i -> let newV = V.drop (i + 1) vals in (True, newV)
```

Comment:
should we be randomly sampling addresses up to the threshold here, or do we want to be on the nose with the hyperlane algorithm?

Reply:
That is apparently the hyperlane algorithm.